static/settings.ts

```diff
@@ -0,0 +1,386 @@
+            .prop('step', sliderSettings.step || 1);
+
+        elem.on('change', this.updateDisplay.bind(this));
+    }
+
+    override putUi(value: number) {
+        this.elem.val(value);
+        this.updateDisplay();
+    }
+
+    override getUi(): number {
+        return parseInt(this.elem.val().toString())
+    }
+
+    private updateDisplay() {
+        this.display.text(this.formatter(this.getUi()));
+    };
+}
+
+class Textbox extends BaseSetting {}
+
+class Numeric extends BaseSetting {
+    private readonly min: number;
+    private readonly max: number;
+
+    constructor(elem: JQuery, name: string, params: Record<'min' | 'max', number>) {
+        super(elem, name);
+
+        this.min = params.min;
+        this.max = params.max;
+
+        elem.attr('min', this.min)
+            .attr('max', this.max);
+    }
+
+    override getUi(): number {
+        return this.clampValue(parseInt(this.elem.val().toString()));
+    }
+
+    override putUi(value: number) {
+        this.elem.val(this.clampValue(value))
+    }
+
+    private clampValue(value: number): number {
+        return Math.min(Math.max(value, this.min), this.max);
+    }
+}
+
+export class Settings {
+    private settingsObjs: BaseSetting[];
+
+    constructor(private root: JQuery,
+                private settings: SiteSettings,
+                private onChange: (SiteSettings) => void,
+                private subLangId: string | null) {
+
+        this.settings = settings;
+        this.settings.defaultLanguage = this.settings.defaultLanguage === null ? undefined : settings.defaultLanguage;
+        this.settingsObjs = [];
+
+        this.addCheckboxes();
+        this.addSelectors();
+        this.addSliders();
+        this.addNumerics();
+        this.addTextBoxes();
+
+        this.setSettings(this.settings);
+        this.handleThemes();
+    }
+
+    public static getStoredSettings(): SiteSettings {
+        return JSON.parse(local.get('settings', '{}'));
+    }
+
+    public setSettings(newSettings: SiteSettings) {
+        this.onSettingsChange(newSettings);
+        this.onChange(newSettings);
+    }
+
+    private onUiChange() {
+        for (let setting of this.settingsObjs) {
+            this.settings[setting.name] = setting.getUi();
+        }
+        this.onChange(this.settings);
+    }
+
+    private onSettingsChange(settings: SiteSettings) {
+        this.settings = settings;
+        for (let setting of this.settingsObjs) {
+            setting.putUi(this.settings[setting.name]);
+        }
+    }
+
+    private add<Type extends BaseSetting>(setting: Type, defaultValue: any) {
+        const key = setting.name;
+        if (this.settings[key] === undefined) this.settings[key] = defaultValue;
+        this.settingsObjs.push(setting);
+        setting.elem.on('change', this.onUiChange.bind(this))
+    }
+
+    private addCheckboxes() {
+        const checkboxes: [string, keyof SiteSettings, boolean][]= [
```

Comment:
Could be nice to have a comment here saying what each array member is, maybe
```suggestion
        // Known checkbox options in order [selector, key, defaultValue]
        const checkboxes: [string, keyof SiteSettings, boolean][]= [
```